File: 02-passwordPhilosophy/partTwo.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

type Password struct {
	plaintext    string
	targetLetter rune
	positionOne  int
	positionTwo  int
}

var parserRegex = regexp.MustCompile(`(?m)(\d+)-(\d+) ([a-z]): (.+)`)

func main() {
	inputBytes, _ := ioutil.ReadFile("input.txt")
	inputSlice := strings.Split(strings.TrimSpace(string(inputBytes)), "\n")

	var passwords []Password
	for _, line := range inputSlice {
		matches := parserRegex.FindAllStringSubmatch(line, -1)
		miR, _ := strconv.Atoi(matches[0][1])
		maR, _ := strconv.Atoi(matches[0][2])
		passwords = append(passwords, Password{
			plaintext:    matches[0][4],
			targetLetter: rune(matches[0][3][0]),
			positionOne:  miR - 1,
			positionTwo:  maR - 1,
		})
	}

	var num_valid_passwords int

	for _, password := range passwords {
		positionOneMatches := rune(password.plaintext[password.positionOne]) == password.targetLetter
		positionTwoMatches := rune(password.plaintext[password.positionTwo]) == password.targetLetter

		if positionOneMatches != positionTwoMatches {
			num_valid_passwords += 1
		}
	}

	fmt.Printf("There are %d valid passwords.\n", num_valid_passwords)
}

File: 02-passwordPhilosophy/python/partTwo.py
import re

from common import *

def partTwo(instr:str) -> int:
    input_string = parse(instr)

    class Password:
        plaintext: str
        target_letter: str
        position_one: int
        position_two: int

        def __init__(self, plaintext:str, target_letter:str, position_one:int, position_two:int) -> None:
            self.plaintext = plaintext
            self.target_letter = target_letter
            self.position_one = position_one - 1  # No concept of index zero... eurgh
            self.position_two = position_two - 1

    parser_regex = r"(\d+)-(\d+) ([a-z]): (.+)"

    passwords = []

    for line in input_string:
        m = re.match(parser_regex, line)
        passwords.append(Password(m.group(4), m.group(3), int(m.group(1)), int(m.group(2))))

    num_valid_passwords = 0

    for password in passwords:
        position_one_matches = password.plaintext[password.position_one] == password.target_letter
        position_two_matches = password.plaintext[password.position_two] == password.target_letter

        if position_one_matches ^ position_two_matches:
            num_valid_passwords += 1        
    
    return num_valid_passwords
